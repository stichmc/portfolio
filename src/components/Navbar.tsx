import * as React from "react";
import { Link } from "react-router-dom";
import { cn } from "@/lib/utils";
import {
  NavigationMenu,
  NavigationMenuContent,
  NavigationMenuItem,
  NavigationMenuLink,
  NavigationMenuList,
  NavigationMenuTrigger,
  navigationMenuTriggerStyle,
} from "@/components/ui/navigation-menu";

const projectComponents: {
  title: string;
  href: string;
  description: string;
}[] = [
  {
    title: "Beyond Game Engine",
    href: "/projects/beyond-engine",
    description: "A 3D video game engine utilizing the Vulkan graphics API.",
  },
  {
    title: "Speech-to-text translator",
    href: "/projects/speech-to-text",
    description: "A real-time deep learning-based speech-to-text translator.",
  },
  {
    title: "Time Escapement",
    href: "/projects/time-escapement",
    description: "A 17th-century time escapement clock.",
  },
  {
    title: "C++ Console Based Game",
    href: "/projects/doom",
    description: "A 2D video game entirely in C++, inspired by DOOM.",
  },
  {
    title: "Wheel Wizard",
    href: "/projects/wheel-wizard",
    description:
      "A used-car website for browsing, posting, and purchasing used cars.",
  },
];

export default function Navbar() {
  return (
    <div className="flex justify-center p-4">
      <NavigationMenu>
        <NavigationMenuList>
          <NavigationMenuItem>
            <Link to="/home">
              <NavigationMenuLink className={navigationMenuTriggerStyle()}>
                <p className="text-lg font-bold">Home</p>
              </NavigationMenuLink>
            </Link>
          </NavigationMenuItem>

          <NavigationMenuItem>
            <Link to="/education">
              <NavigationMenuLink className={navigationMenuTriggerStyle()}>
                <p className="text-lg font-bold">Education</p>
              </NavigationMenuLink>
            </Link>
          </NavigationMenuItem>

          <NavigationMenuItem>
            <Link to="/work-experience">
              <NavigationMenuLink className={navigationMenuTriggerStyle()}>
                <p className="text-lg font-bold">Work Experience</p>
              </NavigationMenuLink>
            </Link>
          </NavigationMenuItem>

          <NavigationMenuItem>
            <NavigationMenuTrigger className="text-lg font-bold">
              Projects
            </NavigationMenuTrigger>
            <NavigationMenuContent>
              <ul className="grid gap-3 p-4 md:w-[400px] lg:w-[500px] lg:grid-cols-[.75fr_1fr]">
                <li className="row-span-3">
                  <NavigationMenuLink asChild>
                    <Link
                      className="flex h-full w-full select-none flex-col justify-end rounded-md bg-gradient-to-b from-muted/50 to-muted p-6 no-underline outline-none focus:shadow-md"
                      to="/projects/synergy"
                    >
                      <img className="h-14 w-14" src="/synergy_logo.svg" />
                      <div className="mb-2 mt-4 text-lg font-medium">
                        Synergy
                      </div>
                      <p className="text-sm leading-tight text-muted-foreground">
                        A more trustworthy, reliable, and secure platform for
                        collaboration and communication.
                      </p>
                    </Link>
                  </NavigationMenuLink>
                </li>
                {projectComponents.map((projectComponent) => (
                  <ListItem
                    key={projectComponent.title}
                    title={projectComponent.title}
                    href={projectComponent.href}
                  >
                    {projectComponent.description}
                  </ListItem>
                ))}
              </ul>
            </NavigationMenuContent>
          </NavigationMenuItem>
        </NavigationMenuList>
      </NavigationMenu>
      <hr className="border-b my-2" />
    </div>
  );
}

const ListItem = React.forwardRef<
  React.ElementRef<"a">,
  React.ComponentPropsWithoutRef<"a">
>(({ className, title, children, ...props }, ref) => {
  return (
    <li>
      <NavigationMenuLink asChild>
        <a
          ref={ref}
          className={cn(
            "block select-none space-y-1 rounded-md p-3 leading-none no-underline outline-none transition-colors hover:bg-accent hover:text-accent-foreground focus:bg-accent focus:text-accent-foreground",
            className
          )}
          {...props}
        >
          <div className="text-sm font-medium leading-none">{title}</div>
          <p className="line-clamp-2 text-sm leading-snug text-muted-foreground">
            {children}
          </p>
        </a>
      </NavigationMenuLink>
    </li>
  );
});
ListItem.displayName = "ListItem";
