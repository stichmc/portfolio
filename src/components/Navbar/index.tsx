import env from "@/env";
import { Button } from "@/components-shadcn/button";
import { cn } from "@/lib/utils";

export interface NavbarProps {
  selectedSection: string;
  handleSetSelectedSection: (section: "about" | "skills" | "work" | "projects") => void;
}
const Navbar = ({ selectedSection, handleSetSelectedSection }: NavbarProps) => {
  return (
    <>
      <div className="fixed h-[80px] top-0 left-0 w-screen flex flex-row items-center justify-between p-2 sm:p-4 md:p-8 lg:p-12 gap-2 z-50 bg-inherit">
        <h3 className="scroll-m-20 text-2xl font-semibold tracking-tight">mcstich.com</h3>

        <div className="flex flex-row gap-2 items-center justify-center">
          <Button
            variant="link"
            className={cn("text-lg select-none", {
              "text-primary": selectedSection === "about",
              "text-primary/75": selectedSection !== "about",
            })}
            onClick={() => handleSetSelectedSection("about")}
          >
            About
          </Button>

          <Button
            variant="link"
            className={cn("text-lg select-none", {
              "text-primary": selectedSection === "skills",
              "text-primary/75": selectedSection !== "skills",
            })}
            onClick={() => handleSetSelectedSection("skills")}
          >
            Skills
          </Button>

          <Button
            variant="link"
            className={cn("text-lg select-none", {
              "text-primary": selectedSection === "work",
              "text-primary/75": selectedSection !== "work",
            })}
            onClick={() => handleSetSelectedSection("work")}
          >
            Work
          </Button>

          <Button
            variant="link"
            className={cn("text-lg select-none", {
              "text-primary": selectedSection === "projects",
              "text-primary/75": selectedSection !== "projects",
            })}
            onClick={() => handleSetSelectedSection("projects")}
          >
            Projects
          </Button>
        </div>

        <Button
          className="select-none"
          onClick={() => window.open(`${env.BASE_URL}/Michael_Stich_Resume.pdf`, "_blank")}
        >
          Download My Resume
        </Button>
      </div>

      {/* Account For Navbar Height */}
      <div className="mt-[80px]" />
    </>
  );
};

export default Navbar;
