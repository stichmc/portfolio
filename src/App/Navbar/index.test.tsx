import { expect, test, vi, beforeEach } from "vitest";
import { render } from "@testing-library/react";
import Navbar, { NavbarProps } from "./";

let props: NavbarProps;

beforeEach(() => {
  props = {
    selectedSection: "about",
    handleSetSelectedSection: vi.fn(),
  };
});

test("Navbar", () => {
  const rendering = render(<Navbar {...props} />);
  expect(rendering).toMatchSnapshot();
});
