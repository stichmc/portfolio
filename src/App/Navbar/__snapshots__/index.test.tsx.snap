// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="fixed h-[80px] top-0 left-0 w-screen flex flex-row items-center justify-between p-2 sm:p-4 md:p-8 lg:p-12 gap-2 z-50 bg-secondary"
      >
        <p
          class="scroll-m-20 text-xl font-semibold tracking-tight"
        >
          mcstich.com
        </p>
        <div
          class="flex flex-row gap-2 items-center justify-center"
        >
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline h-10 px-4 py-2 text-lg select-none text-primary"
          >
            About
          </button>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline h-10 px-4 py-2 text-lg select-none text-primary/75"
          >
            Work
          </button>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline h-10 px-4 py-2 text-lg select-none text-primary/75"
          >
            Skills
          </button>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline h-10 px-4 py-2 text-lg select-none text-primary/75"
          >
            Projects
          </button>
        </div>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 select-none"
        >
          View My Resume
        </button>
      </div>
      <div
        class="h-[80px]"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="fixed h-[80px] top-0 left-0 w-screen flex flex-row items-center justify-between p-2 sm:p-4 md:p-8 lg:p-12 gap-2 z-50 bg-secondary"
    >
      <p
        class="scroll-m-20 text-xl font-semibold tracking-tight"
      >
        mcstich.com
      </p>
      <div
        class="flex flex-row gap-2 items-center justify-center"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline h-10 px-4 py-2 text-lg select-none text-primary"
        >
          About
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline h-10 px-4 py-2 text-lg select-none text-primary/75"
        >
          Work
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline h-10 px-4 py-2 text-lg select-none text-primary/75"
        >
          Skills
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 underline-offset-4 hover:underline h-10 px-4 py-2 text-lg select-none text-primary/75"
        >
          Projects
        </button>
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 select-none"
      >
        View My Resume
      </button>
    </div>
    <div
      class="h-[80px]"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
